![](img/330-banner.png)

# Lecture 10: Regression metrics

UBC, 2023-24

Instructor: Varada Kolhatkar and Andrew Roth

## Imports

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.compose import (
    ColumnTransformer,
    TransformedTargetRegressor,
    make_column_transformer,
)
from sklearn.dummy import DummyRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.impute import SimpleImputer
from sklearn.linear_model import Ridge, RidgeCV
from sklearn.metrics import make_scorer, mean_squared_error, r2_score
from sklearn.model_selection import (
    GridSearchCV,
    cross_val_score,
    cross_validate,
    train_test_split,
)
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler
from sklearn.tree import DecisionTreeRegressor

%matplotlib inline

In [2]:
import warnings

warnings.simplefilter(action="ignore", category=FutureWarning)

### Announcements 

- HW4 is due tonight. 
- HW5 will be posted today. It is a project-type homework assignment and will be due after the midterm.  
- Midterm: Thursday, October 26th, 2023 at 6:00pm (~75 mins long)
    - Checkout the midterm information announcement on Piazza (@407)

## Learning outcomes 

From this lecture, students are expected to be able to:

- Carry out feature transformations on somewhat complicated dataset. 
- Visualize transformed features as a dataframe. 
- Use `Ridge` and `RidgeCV`.
- Explain how `alpha` hyperparameter of `Ridge` relates to the fundamental tradeoff. 
- Explain the effect of `alpha` on the magnitude of the learned coefficients. 
- Examine coefficients of transformed features.  
- Appropriately select a scoring metric given a regression problem.
- Interpret and communicate the meanings of different scoring metrics on regression problems.
    - MSE, RMSE, $R^2$, MAPE
- Apply log-transform on the target values in a regression problem with `TransformedTargetRegressor`.

After carrying out preprocessing, why it's useful to get feature names for transformed features?   

### More comments on tackling class imbalance

- In lecture 9 we talked about a few rather simple approaches to deal with class imbalance. 
- If you have a problem such as fraud detection problem where you want to spot rare events, you can also think of this problem as anomaly detection problem and use different kinds of algorithms such as [isolation forests](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.IsolationForest.html).  
- If you are interested in this area, it might be worth checking out this book on this topic.
[Imbalanced Learning: Foundations, Algorithms, and Applications](https://www.amazon.com/dp/1118074629/ref=as_li_ss_tl?&linkCode=sl1&tag=inspiredalgor-20&linkId=615e87a9105582e292ad2b7e2c7ea339&language=en_US)

```{note} 
When you calculate precision, recall, f1 score, by default only the positive label is evaluated, assuming by default that the positive class is labeled 1. This is configurable through the `pos_label` parameter. 
```

## Dataset [[video](https://youtu.be/lgGTKLwNgkQ)]

In this lecture, we'll be using [Kaggle House Prices dataset](https://www.kaggle.com/c/home-data-for-ml-course/). As usual, to run this notebook you'll need to download the data. For this dataset, train and test have already been separated. We'll be working with the train portion in this lecture. 

In [3]:
df = pd.read_csv("data/housing-kaggle/train.csv")
train_df, test_df = train_test_split(df, test_size=0.10, random_state=123)
train_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
302    303          20       RL        118.0    13704   Pave   NaN      IR1   
767    768          50       RL         75.0    12508   Pave   NaN      IR1   
429    430          20       RL        130.0    11457   Pave   NaN      IR1   
1139  1140          30       RL         98.0     8731   Pave   NaN      IR1   
558    559          60       RL         57.0    21872   Pave   NaN      IR2   

     LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal  \
302          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
767          Lvl    AllPub  ...        0    NaN   NaN        Shed    1300   
429          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
1139         Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
558          HLS    AllPub  ...        0    NaN   NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
302       1   2006        WD         Normal     205000  
767       7   2008        WD         Normal     160000  
429       3   2009        WD         Normal     175000  
1139      5   2007        WD         Normal     144000  
558       8   2008        WD         Normal     175000  

[5 rows x 81 columns]

- The supervised machine learning problem is predicting housing price given features associated with properties. 
- Here, the target is `SalePrice`, which is continuous. So it's a **regression problem** (as opposed to classification).

In [4]:
train_df.shape

(1314, 81)

### Let's separate `X` and `y`

In [5]:
X_train = train_df.drop(columns=["SalePrice"])
y_train = train_df["SalePrice"]

X_test = test_df.drop(columns=["SalePrice"])
y_test = test_df["SalePrice"]

### EDA

In [6]:
train_df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1314.000000  1314.000000  1089.000000    1314.000000  1314.000000   
mean    734.182648    56.472603    69.641873   10273.261035     6.076104   
std     422.224662    42.036646    23.031794    8997.895541     1.392612   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     369.250000    20.000000    59.000000    7500.000000     5.000000   
50%     735.500000    50.000000    69.000000    9391.000000     6.000000   
75%    1099.750000    70.000000    80.000000   11509.000000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1314.000000  1314.000000   1314.000000  1307.000000  1314.000000  ...   
mean      5.570015  1970.995434   1984.659056   102.514155   441.425419  ...   
std       1.112848    30.198127     20.639754   178.301563   459.276687  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1953.000000   1966.250000     0.000000     0.000000  ...   
50%       5.000000  1972.000000   1993.000000     0.000000   376.000000  ...   
75%       6.000000  2000.000000   2004.000000   165.500000   704.750000  ...   
max       9.000000  2010.000000   2010.000000  1378.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1314.000000  1314.000000    1314.000000  1314.000000  1314.000000   
mean     94.281583    45.765601      21.726788     3.624049    13.987062   
std     125.436492    65.757545      60.766423    30.320430    53.854129   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    24.000000       0.000000     0.000000     0.000000   
75%     168.000000    66.750000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1314.000000   1314.000000  1314.000000  1314.000000    1314.000000  
mean      3.065449     46.951294     6.302131  2007.840183  179802.147641  
std      42.341109    522.283421     2.698206     1.332824   79041.260572  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129600.000000  
50%       0.000000      0.000000     6.000000  2008.000000  162000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  212975.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [7]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1314 entries, 302 to 1389
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1314 non-null   int64  
 1   MSSubClass     1314 non-null   int64  
 2   MSZoning       1314 non-null   object 
 3   LotFrontage    1089 non-null   float64
 4   LotArea        1314 non-null   int64  
 5   Street         1314 non-null   object 
 6   Alley          81 non-null     object 
 7   LotShape       1314 non-null   object 
 8   LandContour    1314 non-null   object 
 9   Utilities      1314 non-null   object 
 10  LotConfig      1314 non-null   object 
 11  LandSlope      1314 non-null   object 
 12  Neighborhood   1314 non-null   object 
 13  Condition1     1314 non-null   object 
 14  Condition2     1314 non-null   object 
 15  BldgType       1314 non-null   object 
 16  HouseStyle     1314 non-null   object 
 17  OverallQual    1314 non-null   int64  
 18  OverallCond

### `pandas_profiler`

We do not have `pandas_profiling` in our course environment. You will  have to install it in the environment on your own if you want to run the code below. 

```conda install -c conda-forge pandas-profiling```

In [39]:
data = pd.DataFrame(data = {"a": list(range(0, 10)),"b": list(range(100, 200, 10))})
profile = ProfileReport(df, title="Profiling Report")
profile.to_widgets()

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Unexpected exception formatting exception. Falling back to standard exception


Traceback (most recent call last):
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3526, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "/var/folders/b3/g26r0dcx4b35vf3nk31216hc0000gr/T/ipykernel_10768/69711575.py", line 3, in <module>
    profile.to_widgets()
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 516, in to_widgets
    display(self.widgets)
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 289, in widgets
    isinstance(self.description_set.table["n"], list)
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 253, in description_set
    self._description_set = describe_df(
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/model/describe.py", li

In [32]:
import numpy as np
import pandas as pd
from ydata_profiling import ProfileReport

df = pd.DataFrame(np.random.rand(100, 5), columns=["a", "b", "c", "d", "e"])

In [35]:
profile = ProfileReport(df, title="Profiling Report")
profile.to_widgets()

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Unexpected exception formatting exception. Falling back to standard exception


Traceback (most recent call last):
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3526, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "/var/folders/b3/g26r0dcx4b35vf3nk31216hc0000gr/T/ipykernel_10768/2135232660.py", line 2, in <module>
    profile.to_widgets()
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 516, in to_widgets
    display(self.widgets)
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 289, in widgets
    isinstance(self.description_set.table["n"], list)
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 253, in description_set
    self._description_set = describe_df(
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/model/describe.py", 

In [30]:
data

acousticness  danceability  duration_ms  energy  instrumentalness  key  \
0          0.01020         0.833       204600   0.434          0.021900    2   
1          0.19900         0.743       326933   0.359          0.006110    1   
2          0.03440         0.838       185707   0.412          0.000234    2   
3          0.60400         0.494       199413   0.338          0.510000    5   
4          0.18000         0.678       392893   0.561          0.512000    5   
...            ...           ...          ...     ...               ...  ...   
2012       0.00106         0.584       274404   0.932          0.002690    1   
2013       0.08770         0.894       182182   0.892          0.001670    1   
2014       0.00857         0.637       207200   0.935          0.003990    0   
2015       0.00164         0.557       185600   0.992          0.677000    1   
2016       0.00281         0.446       204520   0.915          0.000039    9   

      liveness  loudness  mode  speechiness    tempo  time_signature  valence  \
0       0.1650    -8.795     1       0.4310  150.062             4.0    0.286   
1       0.1370   -10.401     1       0.0794  160.083             4.0    0.588   
2       0.1590    -7.148     1       0.2890   75.044             4.0    0.173   
3       0.0922   -15.236     1       0.0261   86.468             4.0    0.230   
4       0.4390   -11.648     0       0.0694  174.004             4.0    0.904   
...        ...       ...   ...          ...      ...             ...      ...   
2012    0.1290    -3.501     1       0.3330   74.976             4.0    0.211   
2013    0.0528    -2.663     1       0.1310  110.041             4.0    0.867   
2014    0.2140    -2.467     1       0.1070  150.082             4.0    0.470   
2015    0.0913    -2.735     1       0.1330  150.011             4.0    0.623   
2016    0.2180    -6.221     1       0.1410  190.013             4.0    0.402   

      target                            song_title            artist  
0          1                              Mask Off            Future  
1          1                               Redbone  Childish Gambino  
2          1                          Xanny Family            Future  
3          1                        Master Of None       Beach House  
4          1                        Parallel Lines       Junior Boys  
...      ...                                   ...               ...  
2012       0   Like A Bitch - Kill The Noise Remix    Kill The Noise  
2013       0                                 Candy    Dillon Francis  
2014       0  Habit - Dack Janiels & Wenzday Remix          Rain Man  
2015       0                         First Contact        Twin Moons  
2016       0                    I Wanna Get Better         Bleachers  

[2017 rows x 16 columns]

In [31]:
import pandas as pd
from ydata_profiling import ProfileReport
import matplotlib 

# profile = ProfileReport(data, title="Pandas Profiling Report")  # , minimal=True)
profile = ProfileReport(data, title="Pandas Profiling Report", minimal=True)
profile.to_file("output_report.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Unexpected exception formatting exception. Falling back to standard exception


Traceback (most recent call last):
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3526, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "/var/folders/b3/g26r0dcx4b35vf3nk31216hc0000gr/T/ipykernel_10768/3582944598.py", line 7, in <module>
    profile.to_file("output_report.html")
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 355, in to_file
    data = self.to_html()
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 470, in to_html
    return self.html
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 277, in html
    self._html = self._render_html()
  File "/Users/kvarada/miniconda3/envs/cpsc330/lib/python3.10/site-packages/ydata_profiling/profile_report.py", line 385, in _render_html
    repo

### Feature types 

- Do not blindly trust all the info given to you by automated tools. 
- How does pandas profiling figure out the data type?
    - You can look at the Python data type and say floats are numeric, strings are categorical.
    - However, in doing so you would miss out on various subtleties such as some of the string features being ordinal rather than truly categorical.
    - Also, it will think free text is categorical.

- In addition to tools such as above, it's important to go through data description to understand the data.
- The data description for our dataset is available [here](https://www.kaggle.com/c/home-data-for-ml-course/data?select=data_description.txt).     

### Feature types 

- We have mixed feature types and a bunch of missing values. 
- Now, let's identify feature types and transformations. 

- Let's get the numeric-looking columns. 

In [ ]:
numeric_looking_columns = X_train.select_dtypes(include=np.number).columns.tolist()
print(numeric_looking_columns)

Not all numeric looking columns are necessarily numeric. 

In [ ]:
train_df["MSSubClass"].unique()

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

Also, month sold is more of a categorical feature than a numeric feature. 

In [ ]:
train_df["MoSold"].unique()  # Month Sold

In [ ]:
drop_features = ["Id"]
numeric_features = [
    "BedroomAbvGr",
    "KitchenAbvGr",
    "LotFrontage",
    "LotArea",
    "OverallQual",
    "OverallCond",
    "YearBuilt",
    "YearRemodAdd",
    "MasVnrArea",
    "BsmtFinSF1",
    "BsmtFinSF2",
    "BsmtUnfSF",
    "TotalBsmtSF",
    "1stFlrSF",
    "2ndFlrSF",
    "LowQualFinSF",
    "GrLivArea",
    "BsmtFullBath",
    "BsmtHalfBath",
    "FullBath",
    "HalfBath",
    "TotRmsAbvGrd",
    "Fireplaces",
    "GarageYrBlt",
    "GarageCars",
    "GarageArea",
    "WoodDeckSF",
    "OpenPorchSF",
    "EnclosedPorch",
    "3SsnPorch",
    "ScreenPorch",
    "PoolArea",
    "MiscVal",
    "YrSold",
]

```{note}
I've not looked at all the features carefully. It might be appropriate to apply some other encoding on some of the numeric features above. 
```

In [ ]:
set(numeric_looking_columns) - set(numeric_features) - set(drop_features)

We'll treat the above numeric-looking features as categorical features. 

- There are a bunch of ordinal features in this dataset. 
- Ordinal features with the same scale 
    - Poor (Po), Fair (Fa), Typical (TA), Good (Gd), Excellent (Ex)
    - These we'll be calling `ordinal_features_reg`.
- Ordinal features with different scales
    - These we'll be calling `ordinal_features_oth`.

In [ ]:
ordinal_features_reg = [
    "ExterQual",
    "ExterCond",
    "BsmtQual",
    "BsmtCond",
    "HeatingQC",
    "KitchenQual",
    "FireplaceQu",
    "GarageQual",
    "GarageCond",
    "PoolQC",
]
ordering = [
    "Po",
    "Fa",
    "TA",
    "Gd",
    "Ex",
]  # if N/A it will just impute something, per below
ordering_ordinal_reg = [ordering] * len(ordinal_features_reg)
ordering_ordinal_reg

We'll pass the above as categories in our `OrdinalEncoder`. 

- There are a bunch more ordinal features using different scales.
  - These we'll be calling `ordinal_features_oth`. 
  - We are encoding them separately. 

In [ ]:
ordinal_features_oth = [
    "BsmtExposure",
    "BsmtFinType1",
    "BsmtFinType2",
    "Functional",
    "Fence",
]
ordering_ordinal_oth = [
    ["NA", "No", "Mn", "Av", "Gd"],
    ["NA", "Unf", "LwQ", "Rec", "BLQ", "ALQ", "GLQ"],
    ["NA", "Unf", "LwQ", "Rec", "BLQ", "ALQ", "GLQ"],
    ["Sal", "Sev", "Maj2", "Maj1", "Mod", "Min2", "Min1", "Typ"],
    ["NA", "MnWw", "GdWo", "MnPrv", "GdPrv"],
]

The remaining features are categorical features.

In [ ]:
categorical_features = list(
    set(X_train.columns)
    - set(numeric_features)
    - set(ordinal_features_reg)
    - set(ordinal_features_oth)
    - set(drop_features)
)
categorical_features

- We are not doing it here but we can engineer our own features too. 
- Would price per square foot be a good feature to add in here?

### Applying feature transformations

- Since we have mixed feature types, let's use `ColumnTransformer` to apply different transformations on different features types.  

In [ ]:
from sklearn.compose import make_column_transformer

numeric_transformer = make_pipeline(SimpleImputer(strategy="median"), StandardScaler())
ordinal_transformer_reg = make_pipeline(
    SimpleImputer(strategy="most_frequent"),
    OrdinalEncoder(categories=ordering_ordinal_reg),
)

ordinal_transformer_oth = make_pipeline(
    SimpleImputer(strategy="most_frequent"),
    OrdinalEncoder(categories=ordering_ordinal_oth),
)

categorical_transformer = make_pipeline(
    SimpleImputer(strategy="constant", fill_value="missing"),
    OneHotEncoder(handle_unknown="ignore", sparse_output=False),
)

preprocessor = make_column_transformer(
    ("drop", drop_features),
    (numeric_transformer, numeric_features),
    (ordinal_transformer_reg, ordinal_features_reg),
    (ordinal_transformer_oth, ordinal_features_oth),
    (categorical_transformer, categorical_features),
)

### Examining the preprocessed data

In [ ]:
preprocessor.fit(X_train)  # Calling fit to examine all the transformers.
preprocessor.named_transformers_

In [ ]:
ohe_columns = list(
    preprocessor.named_transformers_["pipeline-4"]
    .named_steps["onehotencoder"]
    .get_feature_names_out(categorical_features)
)
new_columns = (
    numeric_features + ordinal_features_reg + ordinal_features_oth + ohe_columns
)

In [ ]:
X_train_enc = pd.DataFrame(
    preprocessor.transform(X_train), index=X_train.index, columns=new_columns
)
X_train_enc.head()

In [ ]:
X_train.shape

In [ ]:
X_train_enc.shape

We went from 80 features to 263 features!! 

### Other possible preprocessing?  

- There is a lot of room for improvement ...
- We're just using `SimpleImputer`.
    - In reality we'd want to go through this more carefully.
    - We may also want to drop some columns that are almost entirely missing.    
- We could also check for outliers, and do other exploratory data analysis (EDA).
- But for now this is good enough ...    

<br><br><br><br>

## Model building 

### `DummyRegressor`

In [ ]:
dummy = DummyRegressor()
pd.DataFrame(cross_validate(dummy, X_train, y_train, cv=10, return_train_score=True))

### Let's try a linear model: `Ridge`

- Recall that we are going to use `Ridge()` instead of `LinearRegression()` in this course. 
- Similar to linear regression, ridge regression is also a linear model for regression. 
- So the formula it uses to make predictions is the same one used for ordinary least squares. 
- But it has a hyperparameter `alpha` which controls the fundamental tradeoff.     

In [ ]:
lr = make_pipeline(preprocessor, Ridge())
lr.fit(X_train, y_train);

In [ ]:
lr_preds = lr.predict(X_test)
lr_preds[:10]

In [ ]:
lr_preds.max(), lr_preds.min()

In [ ]:
print("Smallest coefficient: ", lr.named_steps["ridge"].coef_.min())
print("Largest coefficient:", lr.named_steps["ridge"].coef_.max())

Let's carry out cross-validation with `Ridge`. 

In [ ]:
lr_pipe = make_pipeline(preprocessor, Ridge())
pd.DataFrame(cross_validate(lr_pipe, X_train, y_train, cv=10, return_train_score=True))

- Quite a bit of variation in the test scores. 
- Performing poorly in fold 8. Not sure why. 
    - Probably it contains the outliers in the data which we kind of ignored. 

### Feature names of transformed data

- If you want to get the column names of newly created columns, you need to fit the transformer. 

In [ ]:
preprocessor

In [ ]:
ohe_columns = list(
    preprocessor.named_transformers_["pipeline-4"]
    .named_steps["onehotencoder"]
    .get_feature_names_out(categorical_features)
)
ohe_columns

In [ ]:
new_columns = (
    numeric_features + ordinal_features_reg + ordinal_features_oth + ohe_columns
)

### Tuning `alpha` hyperparameter of `Ridge`
- Recall that `Ridge` has a hyperparameter `alpha` that controls the fundamental tradeoff.
- This is like `C` in `LogisticRegression` but, annoyingly, `alpha` is the inverse of `C`.
- That is, large `C` is like small `alpha` and vice versa.
- Smaller `alpha`: lower training error (overfitting)

In [ ]:
param_grid = {"ridge__alpha": 10.0 ** np.arange(-5, 5, 1)}

pipe_ridge = make_pipeline(preprocessor, Ridge())

search = GridSearchCV(pipe_ridge, param_grid, return_train_score=True, n_jobs=-1)
search.fit(X_train, y_train)
train_scores = search.cv_results_["mean_train_score"]
cv_scores = search.cv_results_["mean_test_score"]

In [ ]:
plt.semilogx(param_grid["ridge__alpha"], train_scores.tolist(), label="train")
plt.semilogx(param_grid["ridge__alpha"], cv_scores.tolist(), label="cv")
plt.legend()
plt.xlabel("alpha")
plt.ylabel("score");

In [ ]:
best_alpha = search.best_params_
best_alpha

In [ ]:
search.best_score_

- It seems alpha=100 is the best choice here.

- General intuition: larger `alpha` leads to smaller coefficients.
- Smaller coefficients mean the predictions are less sensitive to changes in the data. Hence less chance of overfitting.  

In [ ]:
pipe_bigalpha = make_pipeline(preprocessor, Ridge(alpha=1000))
pipe_bigalpha.fit(X_train, y_train)
bigalpha_coeffs = pipe_bigalpha.named_steps["ridge"].coef_
pd.DataFrame(
    data=bigalpha_coeffs, index=new_columns, columns=["Coefficients"]
).sort_values(by="Coefficients", ascending=False)

- Smaller `alpha` leads to bigger coefficients. 

In [ ]:
pipe_smallalpha = make_pipeline(preprocessor, Ridge(alpha=0.01))
pipe_smallalpha.fit(X_train, y_train)
smallalpha_coeffs = pipe_smallalpha.named_steps["ridge"].coef_
pd.DataFrame(
    data=smallalpha_coeffs, index=new_columns, columns=["Coefficients"]
).sort_values(by="Coefficients", ascending=False)

With the best alpha found by the grid search, the coefficients are somewhere in between. 

In [ ]:
pipe_bestalpha = make_pipeline(
    preprocessor, Ridge(alpha=search.best_params_["ridge__alpha"])
)
pipe_bestalpha.fit(X_train, y_train)
bestalpha_coeffs = pipe_bestalpha.named_steps["ridge"].coef_
pd.DataFrame(
    data=bestalpha_coeffs, index=new_columns, columns=["Coefficients"]
).sort_values(by="Coefficients", ascending=False)

To summarize: 
- Higher values of `alpha` means a more restricted model.  
- The values of coefficients are likely to be smaller for higher values of `alpha` compared to lower values of alpha. 

### `RidgeCV`

Because it's so common to want to tune `alpha` with `Ridge`, sklearn provides a class called `RidgeCV`, which automatically tunes `alpha` based on cross-validation.

In [ ]:
alphas = 10.0 ** np.arange(-6, 6, 1)
ridgecv_pipe = make_pipeline(preprocessor, RidgeCV(alphas=alphas, cv=10))
ridgecv_pipe.fit(X_train, y_train);

In [ ]:
best_alpha = ridgecv_pipe.named_steps["ridgecv"].alpha_

best_alpha

<br><br>

Let's examine the tuned model.

In [ ]:
ridge_tuned = make_pipeline(preprocessor, Ridge(alpha=best_alpha))
ridge_tuned.fit(X_train, y_train)
ridge_preds = ridge_tuned.predict(X_test)
ridge_preds[:10]

In [ ]:
df = pd.DataFrame(
    data={"coefficients": ridge_tuned.named_steps["ridge"].coef_}, index=new_columns
)

In [ ]:
df.sort_values("coefficients", ascending=False)

So according to this model:

- As `OverallQual` feature gets bigger the housing price will get bigger.
- `Neighborhood_Edwards` is associated with reducing the housing price. 
    - We'll talk more about interpretation of different kinds of features next week.

In [ ]:
ridge_preds.max(), ridge_preds.min()

<br><br>

## ❓❓ Questions for you

### iClicker Exercise 10.1

**iClicker cloud join link: https://join.iclicker.com/SNBF**

**Select all of the following statements which are TRUE.**

- (A) Price per square foot would be a good feature to add in our `X`. 
- (B) The `alpha` hyperparameter of `Ridge` has similar interpretation of `C` hyperparameter of `LogisticRegression`; higher `alpha` means more complex model. 
- (C) In `Ridge`, smaller alpha means bigger coefficients whereas bigger alpha means smaller coefficients.  

Can we use the metrics we looked at in the previous lecture for this problem? Why or why not? 

<br><br><br><br>

## Regression scoring functions


- We aren't doing classification anymore, so we can't just check for equality:

In [ ]:
ridge_tuned.predict(X_train) == y_train

In [ ]:
y_train.values

In [ ]:
ridge_tuned.predict(X_train)

We need a score that reflects how right/wrong each prediction is.

There are a number of popular scoring functions for regression. We are going to look at some common metrics: 

- mean squared error (MSE)
- $R^2$
- root mean squared error (RMSE)
- MAPE

See [sklearn documentation](https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics) for more details. 

### Mean squared error (MSE)

- A common metric is mean squared error:

In [ ]:
preds = ridge_tuned.predict(X_train)

In [ ]:
np.mean((y_train - preds) ** 2)

Perfect predictions would have MSE=0.

In [ ]:
np.mean((y_train - y_train) ** 2)

This is also implemented in sklearn:

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_train, preds)

- MSE looks huge and unreasonable. There is an error of ~\$1 Billion!
- Is this score good or bad?

- Unlike classification, with regression **our target has units**. 
- The target is in dollars, the mean squared error is in $dollars^2$ 
- The score also depends on the scale of the targets. 
- If we were working in cents instead of dollars, our MSE would be $10,000 \times (100^2$) higher!

In [ ]:
np.mean((y_train * 100 - preds * 100) ** 2)

### Root mean squared error or RMSE

- The MSE above is in $dollars^2$.
- A more relatable metric would be the root mean squared error, or RMSE

In [ ]:
np.sqrt(mean_squared_error(y_train, ridge_tuned.predict(X_train)))

- Error of \$30,000 makes more sense.
- Let's dig deeper. 

In [ ]:
plt.scatter(y_train, ridge_tuned.predict(X_train), alpha=0.3)
grid = np.linspace(y_train.min(), y_train.max(), 1000)
plt.plot(grid, grid, "--k")
plt.xlabel("true price")
plt.ylabel("predicted price");

- Here we can see a few cases where our prediction is way off.
- Is there something weird about those houses, perhaps? Outliers? 
- Under the line means we're under-predicting, over the line means we're over-predicting.

### $R^2$ (not in detail)

A common score is the $R^2$

- This is the score that `sklearn` uses by default when you call score()
- You can [read about it](https://en.wikipedia.org/wiki/Coefficient_of_determination) if interested.
- $R^2$ measures the proportion of variability in $y$ that can be explained using $X$. 
- Independent of the scale of $y$. So the max is 1.  

$$R^2(y, \hat{y}) = 1 - \frac{\sum_{i=1}^n (y_i - \hat{y_i})^2}{\sum_{i=1}^n (y_i - \bar{y})^2}$$

- The denominator measures the total variance in $y$.  
- The amount of variability that is left unexplained after performing regression.

Key points:
- The maximum is 1 for perfect predictions
- Negative values are very bad: "worse than DummyRegressor" (very bad)


(Optional) Warning: MSE is "reversible" but $R^2$ is not:

In [ ]:
mean_squared_error(y_train, preds)

In [ ]:
mean_squared_error(preds, y_train)

In [ ]:
r2_score(y_train, preds)

In [ ]:
r2_score(preds, y_train)

- When you call `fit` it minimizes MSE / maximizes $R^2$ (or something like that) by default.
- Just like in classification, this isn't always what you want!!

### MAPE

- We got an RMSE of ~$30,000 before. 

Question: Is an error of \$30,000 acceptable?

In [ ]:
np.sqrt(mean_squared_error(y_train, ridge_tuned.predict(X_train)))

- For a house worth \$600k, it seems reasonable! That's 5% error.
- For a house worth \$60k, that is terrible. It's 50% error.

We have both of these cases in our dataset.

In [ ]:
plt.hist(y_train, bins=100);

How about looking at percent error? 

In [ ]:
pred_train = ridge_tuned.predict(X_train)
percent_errors = (pred_train - y_train) / y_train * 100.0
percent_errors

These are both positive (predict too high) and negative (predict too low).

We can look at the absolute percent error:

In [ ]:
np.abs(percent_errors)

And, like MSE, we can take the average over examples. This is called mean absolute percent error (MAPE).

In [ ]:
def my_mape(true, pred):
    return np.mean(np.abs((pred - true) / true))

In [ ]:
my_mape(y_train, pred_train)

Let's use `sklearn` to calculate MAPE. 

In [ ]:
from sklearn.metrics import mean_absolute_percentage_error

In [ ]:
mean_absolute_percentage_error(y_train, pred_train)

- Ok, this is quite interpretable.
- On average, we have around 10% error.

### Transforming the targets

- When you have prices or count data, the target values are skewed. 
- Let's look at our target column. 

In [ ]:
plt.hist(y_train, bins=100);

- A common trick in such cases is applying a log transform on the target column to make it more normal and less skewed.  
- That is, transform $y\rightarrow \log(y)$.
- Linear regression will usually work better on something that looks more normal. 

In [ ]:
plt.hist(np.log10(y_train), bins=100);

We can incorporate this in our pipeline using `sklearn`. 

In [ ]:
from sklearn.compose import TransformedTargetRegressor

In [ ]:
ttr = TransformedTargetRegressor(
    Ridge(alpha=best_alpha), func=np.log1p, inverse_func=np.expm1
) # transformer for log transforming the target
ttr_pipe = make_pipeline(preprocessor, ttr)

In [ ]:
ttr_pipe

Why can't we incorporate preprocessing targets in our column transformer? 

In [ ]:
ttr_pipe.fit(X_train, y_train); # y_train automatically transformed

In [ ]:
ttr_pipe.predict(X_train)  # predictions automatically un-transformed

In [ ]:
mean_absolute_percentage_error(y_train, ttr_pipe.predict(X_train))

In [ ]:
mean_absolute_percentage_error(y_test, ttr_pipe.predict(X_test))

We reduced MAPE from ~10% to ~8% with this trick! 

- Does `.fit()` know we care about MAPE?
- No, it doesn't. Why are we minimizing MSE (or something similar) if we care about MAPE??
- When minimizing MSE, the expensive houses will dominate because they have the biggest error.

### Different scoring functions with `cross_validate`

- Let's try using MSE instead of the default $R^2$ score. 

In [ ]:
pd.DataFrame(
    cross_validate(
        ridge_tuned,
        X_train,
        y_train,
        return_train_score=True,
        scoring="neg_mean_squared_error",
    )
)

In [ ]:
# make a scorer function that we can pass into cross-validation
mape_scorer = make_scorer(my_mape, greater_is_better=False)

pd.DataFrame(
    cross_validate(
        ridge_tuned, X_train, y_train, return_train_score=True, scoring=mape_scorer
    )
)

If you are finding `greater_is_better=False` argument confusing, here is the documentation: 

> greater_is_better(bool), default=True
Whether score_func is a score function (default), meaning high is good, or a loss function, meaning low is good. In the latter case, the scorer object will sign-flip the outcome of the score_func.

Since our custom scorer `mape` gives an error and not a score, I'm passing `False` to it and it'll sign flip so that we can interpret bigger numbers as better performance.  

In [ ]:
# ?make_scorer

In [ ]:
scoring = {
    "r2": "r2",
    "mape_scorer": mape_scorer, # just for demonstration for a custom scorer
    "sklearn MAPE": "neg_mean_absolute_percentage_error",
    "neg_root_mean_square_error": "neg_root_mean_squared_error",
    "neg_mean_squared_error": "neg_mean_squared_error",
}

pd.DataFrame(
    cross_validate(
        ridge_tuned, X_train, y_train, return_train_score=True, scoring=scoring
    )
).T

Are we getting the same `alpha` with mape? 

In [ ]:
param_grid = {"ridge__alpha": 10.0 ** np.arange(-6, 6, 1)}

pipe_ridge = make_pipeline(preprocessor, Ridge())

search = GridSearchCV(
    pipe_ridge, param_grid, return_train_score=True, n_jobs=-1, scoring=mape_scorer
)
search.fit(X_train, y_train);

In [ ]:
print("Best hyperparameter values: ", search.best_params_)
print("Best score: %0.3f" % (search.best_score_))
pd.DataFrame(search.cv_results_)[
    [
        "mean_train_score",
        "mean_test_score",
        "param_ridge__alpha",
        "mean_fit_time",
        "rank_test_score",
    ]
].set_index("rank_test_score").sort_index().T

### Using multiple metrics in `GridSearchCV` or `RandomizedSearchCV` 

- We could use multiple metrics with `GridSearchCV` or `RandomizedSearchCV`. 
- But if you do so, you need to set `refit` to the metric (string) for which the `best_params_` will be found and used to build the `best_estimator_` on the whole dataset. 

In [ ]:
search_multi = GridSearchCV(
    pipe_ridge,
    param_grid,
    return_train_score=True,
    n_jobs=-1,
    scoring=scoring,
    refit="sklearn MAPE",
)
search_multi.fit(X_train, y_train);

In [ ]:
print("Best hyperparameter values: ", search_multi.best_params_)
print("Best score: %0.3f" % (search_multi.best_score_))
pd.DataFrame(search_multi.cv_results_).set_index("rank_test_mape_scorer").sort_index()

What's the test score? 

In [ ]:
search_multi.score(X_test, y_test)

In [ ]:
my_mape(y_test, ridge_tuned.predict(X_test))

### Using regression metrics with `scikit-learn`

- In `sklearn`, you will notice that it has negative version of the metrics above (e.g., `neg_mean_squared_error`, `neg_root_mean_squared_error`). 
- The reason for this is that scores return a value to maximize, the higher the better.

## ❓❓ Questions for you

### iClicker Exercise 10.2

**iClicker cloud join link: https://join.iclicker.com/SNBF**

**Select all of the following statements which are TRUE.**

- (A) We can use still use precision and recall for regression problems but now we have other metrics we can use as well.
- (B) In `sklearn` for regression problems, using `r2_score()` and `.score()` (with default values) will produce the same results.
- (C) RMSE is always going to be non-negative.
- (D) MSE does not directly provide the information about whether the model is underpredicting or overpredicting.
- (E) We can pass multiple scoring metrics to `GridSearchCV` or `RandomizedSearchCV` for regression as well as classification problems. 


<br><br><br><br>

## What did we learn today? 

- House prices dataset target is price, which is numeric -> regression rather than classification
- There are corresponding versions of all the tools we used:
    - `DummyClassifier` -> `DummyRegressor`
    - `LogisticRegression` -> `Ridge`
- `Ridge` hyperparameter `alpha` is like `LogisticRegression` hyperparameter `C`, but opposite meaning
- We'll avoid `LinearRegression` in this course.

- Scoring metrics
- $R^2$ is the default .score(), it is unitless, 0 is bad, 1 is best
- MSE (mean squared error) is in units of target squared, hard to interpret; 0 is best
- RMSE (root mean squared error) is in the same units as the target; 0 is best
- MAPE (mean absolute percent error) is unitless; 0 is best, 1 is bad

![](img/eva-seeyou.png)